RFC: Performance WG (#74)
django/dsf-working-groups: draft/performance.md

# Django Performance WG

## Scope of responsibilities

> The performance working group is responsible for improving the performance of the Django project, including the main `django/django` codebase and other packages under the Django project umbrella. Additionally, they act as an advisory team to other groups both within the Django project and wider community.

Reviewer: I wonder if we should set expectations somewhere with a brief statement that performance is not the paramount goal. I'll gladly pay a few nanoseconds for more idiomatic, maintainable python.

Author: Do you think this is already covered by:

> The pursuit of performance must never be at the cost of Django's core values, such as stability, ease of use, maintainability and developer experience.

Reviewer: Haha yes 🤦‍♂️

The duties of the working group are:

- Review performance-related PRs across first-party Django projects
- Assisting the Triage and Review team with benchmarks for relevant tickets
- Provide performance guidance on Django features and APIs
- Help maintain Django's performance documentation and best-practice guidance
- Assist with development and maintenance of performance-related packages in the Django ecosystem
- Provide performance advice to Django package maintainers when requested
- Mentor new contributors interested in Django performance

> - Advise community members in improving the performance of their Django applications

Reviewer 2: ```suggestion
- Advise community members in improving the performance of Django's community applications
```
Advice for personal or working projects can be a lot of effort, I tried to leave the point but limiting only for Django's community application.

Feel free to update or remove with an explanation.

Author: I agree in principle - I don't want this to become a consulting arm of the DSF!

However, improving documentation, writing blog posts etc arguably does help the community improve the performance of proprietary applications - and I'd like to support that (where possible).

- Run or support performance-focused sessions in DjangoCon sprints

> - Maintenance of the existng `django-asv` benchmark suite

Reviewer 2: ```suggestion
- Maintenance and upgrade of the existng `django-asv` benchmark suite
```

Author: What do you see as the difference between "maintaining" and "upgrading"?

- Act as an advisory board to the Steering Council when deciding on performance-related enhancements

Performance-related security issues, such as time-based DoS, are not in scope, and are instead handled by the Security Team in the first instance (as currently). However, the Performance WG are available for assistance (triage, benchmark development) if necessary, however this is not a requirement. If the Performance WG discover a performance-related security issue (either themselves or through an issue report), they must report it to the Security Team through existing processes.

Changes to Django and related packages proposed by the Performance WG follow the standard processes for changes, such as DEPs or the `new-features` repository. Final say for inclusion remains with the Steering Council, Merger team or Fellows as relevant. The pursuit of performance must never be at the cost of Django's core values, such as stability, ease of use, maintainability and developer experience.

## Initial membership

- Chair:
- Co-Chair:
- Board Liaison (must be an active Board member; may be the same as Chair/Co-Chair):

> - Other members:

Reviewer 3: ```suggestion
- Other members:
  - Farhan Ali Raza 
```
cc @FarhanAliRaza

## Future membership

### Who is eligible to join?

> Any DSF member is eligible to join. The only requirement is that they bring something to the team. Either experience optimising large codebases, working at large scale with Python or Django, or experience maintaining performance-critical components for / of Django (for example WSGI/ASGI servers, DDT, profilers).

Reviewer 2: I would also add experience with Django's 3rd part packages performance (e.g. psycopg, redis, goal,...) or performance and optimization experience in Python language and it's standard library.

Because of the nature of performance investigations, both within Django and the wider Python ecosystem, it's expected joining this working group may be in high demand. For that reason, once the working group is at a reasonable capacity it may not be open for new formal members. The form will remain open to act as an expression of interest.

To mitigate people engaged community members feeling "left out", the vast majority of discussions should be happening in the open (see [Communication](#communication) below). This enables anyone to contribute to discussions and suggest improvements, without needing to be a formal member of the group.

To avoid unnecessary member churn, WG members are expected to sit for at least 1 release cycle (around 9 months). As a volunteer group, this is not a strict requirement.

> ### How do people who want to join sign up / volunteer / express interest?

Reviewer 2: Can we ask for a PR in this repository as per other WG? Instead of a form in an external service we can add a Template PR with predefined questions and handle everything here.

- A public `#performance` channel in the Django Discord
- A new "Performance" topic under "Django Internals" on the Django Forum
- A `performance-wg` group on GitHub, for convenient mentioning

Meetings: A monthly catch up for the team to review and triage incoming tickets which require further discussion and consider ideas from the group and wider community. These meetings are internal to the group (invite only), however minutes will be published publicly shortly afterwards. The Steering Council and Fellows are considered to have a standing invitation.

Whilst synchronous communication is often needed for larger discussions, asynchronous public discussions should be preferred where possible.

## Reporting

The working group will provide quarterly reports to the DSF board summarizing activities in the current quarter and goals for the upcoming quarter.

## Appendix: What is "Performance"?

"Performance" is a loaded term by design. In this context, it refers to:

- HTTP request / response times
- ORM query efficiency
- Cache efficacy
- Idle / load resource (CPU, memory, IO) usage, both client and server side
- [Core Web Vitals](https://web.dev/articles/vitals)

> - Import time / start-up latency

Reviewer: With lazy imports coming in Python 3.15, I asked a bunch of people about this at PyCon and couldn't understand what the production use case would be.[^1] I suppose having a new container come online to relieve traffic slightly faster at the cost of an unlucky request later paying a penalty. Something like that?

I get the use case for the autoreload devserver. Anyway -- this comment is just hand-in-glove with my other comment: without any disclaimer above, this might come off as a list of "goals" (and I'm not sure faster import time should be one), so if we just add a disclaimer somewhere to set expectations, I'm good 👍 

[^1]:Now I remember asking Jake as well, and he mentioned health checks.

Author: There's still the potential for wins here, even with lazy imports. Django does a fair bit of importing upfront, which could be saved.

I've also had conversations about lazy imports, and can't really see the draw for production (and arguably see value in the opposite). However, there are probably ways to improve the import time without just making things lazy - that's what I'd be most interested exploring.
